modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -512,80 +504,81 @@ private boolean affinityInfoIsUpToDate(int cacheId) {
         return ranges.stream()
             .flatMap(r -> IntStream
                 .rangeClosed(r.portFrom(), r.portTo()).boxed()
-                .map(p -> new InetSocketAddress(r.host(), p))
+                .map(p -> InetSocketAddress.createUnresolved(r.host(), p))
             )
-            .collect(Collectors.toList());
+            .collect(Collectors.toMap(a -> a, a -> 1, Integer::sum));
     }
 
-    /** */
-    private synchronized ClientChannel channel() {
-        if (closed)
-            throw new ClientException("Channel is closed");
-
+    /**
+     * Roll current default channel if specified holder equals to it.
+     */
+    private void rollCurrentChannel(ClientChannelHolder hld) {
+        curChannelsGuard.writeLock().lock();
         try {
-            return channels[curChIdx].getOrCreateChannel();
-        }
-        catch (ClientConnectionException e) {
-            rollCurrentChannel();
+            int idx = curChIdx;
+            List<ClientChannelHolder> holders = channels;
 
-            throw e;
-        }
-    }
+            ClientChannelHolder dfltHld = holders.get(idx);
 
-    /** */
-    private synchronized void rollCurrentChannel() {
-        if (++curChIdx >= channels.length)
-            curChIdx = 0;
+            if (dfltHld == hld) {
+                idx += 1;
+
+                if (idx >= holders.size())
+                    curChIdx = 0;
+                else
+                    curChIdx = idx;
+            }
+        } finally {
+            curChannelsGuard.writeLock().unlock();
+        }
     }
 
     /**
      * On current channel failure.
      */
-    private synchronized void onChannelFailure(ClientChannel ch) {
+    private void onChannelFailure(ClientChannel ch) {
         // There is nothing wrong if curChIdx was concurrently changed, since channel was closed by another thread
         // when current index was changed and no other wrong channel will be closed by current thread because
         // onChannelFailure checks channel binded to the holder before closing it.
-        onChannelFailure(channels[curChIdx], ch);
-
-        chFailLsnrs.forEach(Runnable::run);
+        onChannelFailure(channels.get(curChIdx), ch);
     }
 
     /**
      * On channel of the specified holder failure.
      */
-    private synchronized void onChannelFailure(ClientChannelHolder hld, ClientChannel ch) {
-        if (ch == hld.ch && ch != null) {
+    private void onChannelFailure(ClientChannelHolder hld, ClientChannel ch) {
+        if (ch != null && ch == hld.ch)
             hld.closeChannel();
 
-            if (hld == channels[curChIdx])
-                rollCurrentChannel();
-        }
+        chFailLsnrs.forEach(Runnable::run);
+
+        if (scheduledChannelsReinit.get())
+            channelsInit(true);
+        else
+            rollCurrentChannel(hld);
```

Comment:
Why do we skip rollCurrentChannel if scheduledChannelsReinit? I think we should roll channel in any case.

Reply:
if channel fails and there is scheduled reinit, there are chances that reinitialization will fix the issue (reinitialization may assign new default channel).

Reply:
There are chances if some dynamic address finder is used, but by default, you just check current address twice and skip one channel. In both cases (dynamic and static address finder) there is nothing bad happens if we roll current channel

Reply:
Also, if new default channel is assigned by `channelsInit()`, than `rollCurrentChannel(hld)` will do no-op, since `hld` already not a default channel.

Reply:
Actually there is an invokation sequence for default case: `channelsInit -> applyOnDefaultChannel` to establish connection with default channel (that checks that connection is OK). So rolling is not required too. Also rolling at least acquires lock so it is not free.

So, there is a matrix of cases for channel failure:
1. Fixed addreses, topology changes -> Channel is rolled with `channelsInit -> applyOnDefaultChannel`;
2. Fixed addresses, topology does not change -> `rollCurrentChannel`;
3. Dynamic address, topology changes -> channels are reinited twice, rolls with `applyOnDefaultChannel`;
4. Dynamic address, topology does not change -> `rollCurrentChannel`.

So, there is an issue with double initialization for dynamic addresses only. I will fix it by skipping second initialization as describe in other comment.